core/jvm-native/src/main/scala/zio/ZIOAppPlatformSpecific.scala

```diff
@@ -23,42 +23,62 @@ private[zio] trait ZIOAppPlatformSpecific { self: ZIOApp =>
         result  <- runtime.run(ZIO.scoped[Environment with ZIOAppArgs](run)).tapErrorCause(ZIO.logErrorCause(_))
       } yield result).provideLayer(newLayer.tapErrorCause(ZIO.logErrorCause(_)))
 
-    runtime.unsafe.run {
-      ZIO.uninterruptible {
-        for {
-          fiberId <- ZIO.fiberId
-          fiber <- workflow.interruptible.exitWith { exit0 =>
-                     val exitCode = if (exit0.isSuccess) ExitCode.success else ExitCode.failure
-                     interruptRootFibers(fiberId).as(exitCode)
-                   }.fork
-          _ <-
-            ZIO.succeed(Platform.addShutdownHook { () =>
-              if (shuttingDown.compareAndSet(false, true)) {
+    val shutdownLatch = internal.OneShot.make[Unit]
 
-                if (FiberRuntime.catastrophicFailure.get) {
-                  println(
-                    "**** WARNING ****\n" +
-                      "Catastrophic error encountered. " +
-                      "Application not safely interrupted. " +
-                      "Resources may be leaked. " +
-                      "Check the logs for more details and consider overriding `Runtime.reportFatal` to capture context."
-                  )
-                } else {
-                  // NOTE: try-catch likely not needed,
-                  // but guarding against cases where the submission of the task fails spuriously
-                  try {
-                    fiber.tellInterrupt(Cause.interrupt(fiberId))
-                  } catch {
-                    case _: Throwable =>
-                  }
-                }
+    def shutdownHook(fiberId: FiberId, fiber: Fiber.Runtime[Nothing, ExitCode]): Unit =
+      Platform.addShutdownHook { () =>
+        if (shuttingDown.compareAndSet(false, true)) {
+          if (FiberRuntime.catastrophicFailure.get) {
+            println(
+              "**** WARNING ****\n" +
+                "Catastrophic error encountered. " +
+                "Application not safely interrupted. " +
+                "Resources may be leaked. " +
+                "Check the logs for more details and consider overriding `Runtime.reportFatal` to capture context."
+            )
+          } else {
+            try {
+              fiber.tellInterrupt(Cause.interrupt(fiberId))
+              gracefulShutdownTimeout match {
+                case Duration.Infinity       => shutdownLatch.get()
+                case d if d <= Duration.Zero => ()
+                case d                       => shutdownLatch.get(d.toMillis)
               }
-            })
-          result <- fiber.join
-          _      <- exit(result)
-        } yield ()
+            } catch {
+              case _: OneShot.TimeoutException =>
+                println(
+                  "**** WARNING ****\n" +
+                    s"Timed out waiting for ZIO application to shut down after $gracefulShutdownTimeout. " +
```

Comment:
minor: I would use a multi-line string + `.render` to make the output human readable.
```suggestion
                    s"Timed out waiting for ZIO application to shut down after ${gracefulShutdownTimeout.render}. " +
```